Pwr 646 create sample project in line chatbot

ZEZAY opened the pull request.

ZEZAY committed:
feat(core/plugin): add logger plugin

ZEZAY committed:
feat(core/plugin): update dotenv plugin

ZEZAY committed:
feat(webhook): add webhook domain

ZEZAY committed:
feat(webhook): implement webhook

ZEZAY committed:
feat(messaging): add messaging domain

ZEZAY committed:
feat(messaging): implement messaging

ZEZAY committed:
refactor(webhook): source types

ZEZAY committed:
refactor(messaging): rename type MessagePayload

ZEZAY committed:
feat(messaging): add sendDirectMessage usecase

ZEZAY committed:
refactor(messaging): split dto and implement typebox

ZEZAY committed:
feat(core): implement messaging

ZEZAY committed:
refactor(domain): remove webhook-controller.ts

ZEZAY committed:
refactor(domain): remove empty files

ZEZAY committed:
refactor: split server from index

ZEZAY committed:
refactor(webhook): split dto and implement typebox

ZEZAY committed:
refactor: remove unuses

ZEZAY committed:
feat(mongodb): add domain/mongodb

ZEZAY committed:
feat(mongodb): implement mongodb

ZEZAY committed:
fix(mongodb): return Promise

suthithamk commented on src/core/app-route.ts:
```diff
@@ -0,0 +1,33 @@
+import { Logger } from './plugin';
+
+export async function LineChatbotRoute(fastify: FastifyInstance): Promise<void> {
+  const lineApiUsecase = new LineChatbotUsecase();
```
naming ไม่ consistance

suthithamk commented on src/server.ts:
```diff
@@ -0,0 +1,41 @@
+  Container.set('ChannelAccessToken', channelAccessToken);
+
+  const messenger = new Messenger();
+  Container.set('Messenger', messenger);
```
ตอน set ใช้ class instance ไปเลย จะได้ได้ type ติดไปด้วยตอน get

suthithamk commented on src/domain/messaging/messaging-entity.ts:
สิ่งนี้ดูจากการใช้งานไม่ใช่ entity น่าจะเป็น service(level repository ตาม clean ของฝั่ง external data) มากกว่า

suthithamk commented on src/core/app-usecase.ts:
ไฟล์นี้เหมือนเอา usecase ของ 2 domain มารวมกัน ซึ่งจริงๆแล้ว 2 domain นี้ไม่มีอะไรเกี่ยวกันที่ทำให้ต้องมารวมกันเลย แยกเป็น 2 usecase file ข้างใน domain ไปเลยดีกว่า

suthithamk commented on src/domain/messaging/messaging-dto/index.ts:
naming ค่อนข้างไม่สื่อ ว่า 2 file นี้เก็บอะไร ต้องแยกทำไม

suthithamk commented on src/server.ts:
```diff
@@ -0,0 +1,41 @@
+
+  await connectToDatabase(config.MONGODB_URI, config.MONGODB_DATABASE, config.MONGODB_COLLECTION);
+  const mongoDBUsecase = new MongoDBUsecase();
+  const channelAccessToken = await mongoDBUsecase.getAccessToken(config.CHANNEL_ID);
```
ไม่แน่ใจ scope ของ project นี้เท่าไหร่ว่าต้องการแบบนี้ไหม
แต่ถ้าใช้ท่านี้จะ hard code ว่าต่อได้ channel เดียว

suthithamk commented on src/domain/mongodb/mongodb-usecase.ts:
เหมือนกัน อันนี้หน้าตาเหมือน repository มากกว่า usecase

suthithamk commented on src/domain/mongodb/mongodb-usecase.ts:
```diff
@@ -0,0 +1,64 @@
+import { Logger } from '../../core/plugin';
+import { AccessTokenRecord } from './mongodb-dto';
+
+export async function connectToDatabase(database_uri: string, database_name: string, collection_name: string) {
```
อันนี้แยกไปใส่ใน plugin ดีกว่า

suthithamk commented on src/domain/mongodb/index.ts:
ชื่อ domain ไม่สื่อว่าเอาไว้ใช้ทำอะไร
เราควรตั้งชื่อตาม business logic ที่มันทำ

suthithamk commented on src/domain/webhook/webhook-usecase.ts:
```diff
@@ -0,0 +1,53 @@
+export class WebhookUsecase {
+  async handleWebhookEvent(
+    event: WebhookEvent,
+    logger: Logger = Container.get('Logger'),
```
ไป get ใน contructor

ZEZAY committed:
deploy: add Dockerfile

ZEZAY committed:
fix(core): naming consistency

ZEZAY committed:
refactor: instance LoggerContainerKey

ZEZAY committed:
refactor: instance DotEnvConfig

ZEZAY committed:
refactor: instance AccessTokenRecordContainerKey

ZEZAY committed:
refactor: instance Messenger

ZEZAY committed:
refactor: instance WebhookUsecase

ZEZAY committed:
refactor: instance mongoDB

ZEZAY committed:
refactor(messaging): rename file

ZEZAY committed:
refactor: messaging-dto

ZEZAY committed:
refactor: webhook-dto

ZEZAY committed:
refactor(messaging): remove index.ts

ZEZAY committed:
refactor(mongodb): remove index.ts.

ZEZAY committed:
refactor(webhook): remove index.ts.

ZEZAY committed:
refactor(plugin): move connectToDatabase to plugin

ZEZAY committed:
refactor(mongodb): rename file, class

ZEZAY committed:
refactor(webhook): add WebhookUsecase constructor

ZEZAY committed:
refactor(core): move errorHandler to plugin

ZEZAY committed:
refactor(core): remove app-usecase.ts

ZEZAY committed:
refactor(core): rename file

ZEZAY committed:
chore: fix import, remove todo

suthithamk commented on src/domain/messaging/messaging-service.ts:
```diff
@@ -0,0 +1,53 @@
+  ) {}
+
+  async sendBroadcast(payload: ReplyPayload) {
+    const response = await axios.post(this.baseUrl + '/reply', payload, this.requestConfig).catch(error => {
```
ของสลับกันป่าว

suthithamk commented on src/core/route.ts:
```diff
@@ -0,0 +1,44 @@
+    async (req, res) => {
+      logger.info(`send a direct message`);
+      const messagePayload: DirectMessagePayload = req.body;
+      await messenger.sendDirectMessage(messagePayload).catch(err => {
```
เอา controller ข้ามไปเรียก service เลยไม่ค่อยดีเท่าไหร่ อยากให้มี usecase คั่นนิดนึง

suthithamk commented on src/domain/webhook/webhook-usecase.ts:
```diff
@@ -0,0 +1,63 @@
+        },
+      ],
+    };
+    await this.messenger.sendReply(reply).catch(error => {
```
ทำ catch ตรงนี้ก็ไม่ throw ไปถึง error handling ป่ะนะ
เวลา reply error ก็จะได้ 200

suthithamk commented on src/core/plugin/database-connect.ts:
```diff
@@ -0,0 +1,14 @@
+  await client.connect();
+
+  const db: mongoDB.Db = client.db(database_name);
+  const collection: mongoDB.Collection = db.collection(collection_name);
```
ทำท่านี้คือต่อได้ collection เดียวทั้ง project เลยนะ
ไปเลือก collection ตรง repo ดีกว่า

suthithamk commented on src/domain/mongodb/mongodb-repository.ts:
```diff
@@ -0,0 +1,53 @@
+import { LoggerContainerKey } from '../../core/plugin';
+import { AccessTokenRecord } from './mongodb-dto';
+
+export class MongoDBRepository {
```
ย้ำอีกรอบว่าชื่อ domain นี้ไม่สื่อ
คนมาทำต่อจะไม่รู้ว่าสิ่งนี้เอาไว้ทำอะไร รู้แค่ว่าใช้ mongo

suthithamk commented on src/server.ts:
```diff
@@ -0,0 +1,41 @@
+
+  await connectToDatabase(config.MONGODB_URI, config.MONGODB_DATABASE, config.MONGODB_COLLECTION);
+  const mongoDBRepository = new MongoDBRepository();
+  const channelAccessToken = await mongoDBRepository.getAccessToken(config.CHANNEL_ID);
```
ควรใช้ channel id จาก data จริงมากกว่าจากใน config

ZEZAY committed:
refactor: rename domain/access-token

ZEZAY committed:
refactor: rename domain/access-token

ZEZAY committed:
refactor(access-token)

- rename plugin
- remove collection name from .env
- change repo constructor
- rename func

ZEZAY committed:
fix(messaging): remove error throw

ZEZAY committed:
refactor(messaging): add usecase

ZEZAY committed:
fix(webhook): remove catch error

ZEZAY committed:
refactor(messaging): rename class

ZEZAY committed:
fix: logging

ZEZAY committed:
feat: remove hardcode 'channelId'

ZEZAY committed:
refactor(core/route): move AccessTokenRepository to usecase

ZEZAY committed:
test: add tests

ZEZAY committed:
doc: add readme
